spring-integration-core/src/main/java/org/springframework/integration/transformer/FromProtobufTransformer.java

```diff
@@ -0,0 +1,173 @@
+		assertExpressionNotNull(expression);
+		this.typeIdExpression = expression;
+		return this;
+	}
+
+	/**
+	 * Set the expression to evaluate against the message to determine the type id. Default
+	 * {@code headers['proto_type']}.
+	 * @param expression the expression.
+	 * @return the transformer
+	 */
+	public FromProtobufTransformer typeExpression(String expression) {
+		assertExpressionNotNull(expression);
+		this.typeIdExpression = EXPRESSION_PARSER.parseExpression(expression);
+		return this;
+	}
+
+	/**
+	 * Set the expression to evaluate against the message to determine the type. Default {@code headers['proto_type']}.
+	 * @param expression the expression.
+	 */
+	public void setTypeExpression(Expression expression) {
+		assertExpressionNotNull(expression);
+		this.typeIdExpression = expression;
+	}
+
+	/**
+	 * Set the expression to evaluate against the message to determine the type id. Default
+	 * {@code headers['proto_type']}.
+	 * @param expression the expression.
+	 */
+	public void setTypeExpressionString(String expression) {
+		assertExpressionNotNull(expression);
+		this.typeIdExpression = EXPRESSION_PARSER.parseExpression(expression);
+	}
+
+	private void assertExpressionNotNull(Object expression) {
+		Assert.notNull(expression, "'expression' must not be null");
+	}
+
+	@Override
+	protected void onInit() {
+		this.evaluationContext = IntegrationContextUtils.getEvaluationContext(getBeanFactory());
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	protected Object doTransform(Message<?> message) {
+		Assert.state(message.getPayload() instanceof byte[], "Payload must be a byte[]");
+		Class<? extends com.google.protobuf.Message> type = null;
+		Object value = this.typeIdExpression.getValue(this.evaluationContext, message);
+		if (value instanceof Class) {
+			type = (Class<? extends com.google.protobuf.Message>) value;
+		}
+		else if (value instanceof String) {
+			try {
+				type = (Class<? extends com.google.protobuf.Message>) ClassUtils.forName((String) value,
+						this.beanClassLoader);
+			}
+			catch (ClassNotFoundException | LinkageError e) {
+				throw new IllegalStateException(e);
+			}
+		}
+		if (type == null) {
+			type = this.defaultType;
+		}
+
+		try {
+			Builder messageBuilder = this.getMessageBuilder(type);
+			byte[] payload = (byte[]) message.getPayload();
+			com.google.protobuf.Message boza = messageBuilder.mergeFrom(payload).build();
+			return boza;
+		}
+		catch (IOException e) {
+			throw new UncheckedIOException(e);
+		}
+	}
+
+	/**
+	 * Create a new {@code Message.Builder} instance for the given class.
+	 * <p>
+	 * This method uses a ConcurrentReferenceHashMap for caching method lookups.
+	 */
+	private com.google.protobuf.Message.Builder getMessageBuilder(Class<? extends com.google.protobuf.Message> clazz) {
+		try {
+			Method method = methodCache.get(clazz);
+			if (method == null) {
+				method = clazz.getMethod("newBuilder");
+				methodCache.put(clazz, method);
```

Comment:
`computeIfAbsent()` instead?
And why do use use reflection for this method at all?
Why Protobuf does not have an API for `Message.newBuilder()` ?